pydantic/main.py

```diff
@@ -1216,6 +1237,49 @@ def _calculate_keys(self, *args: Any, **kwargs: Any) -> Any:
         return _deprecated_copy_internals._calculate_keys(self, *args, **kwargs)  # type: ignore
 
 
+RootModelRootType = typing.TypeVar('RootModelRootType')
+
+
+class RootModel(BaseModel, typing.Generic[RootModelRootType]):
+    __pydantic_root_model__ = True
+
+    root: RootModelRootType
+
+    def __init__(__pydantic_self__, data: RootModelRootType) -> None:  # type: ignore
+        __tracebackhide__ = True
+        __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
+
+    __init__.__pydantic_base_init__ = True  # type: ignore
+
+    def model_dump(
+        self,
+        *,
+        mode: str = 'python',
+        include: IncEx = None,
+        exclude: IncEx = None,
+        by_alias: bool = False,
+        exclude_unset: bool = False,
+        exclude_defaults: bool = False,
+        exclude_none: bool = False,
+        round_trip: bool = False,
+        warnings: bool = True,
+    ) -> dict[str, Any]:
+        return super().model_dump(
+            mode=mode,
+            include=include,
+            exclude=exclude,
+            by_alias=by_alias,
+            exclude_unset=exclude_unset,
+            exclude_defaults=exclude_defaults,
+            exclude_none=exclude_none,
+            round_trip=round_trip,
+            warnings=warnings,
+        )['root']
+
+    def __iter__(self) -> TupleGenerator:
+        yield ('__root__', self.model_dump())
```

Comment:
```suggestion
        yield ('root', self.model_dump())
```
Thinking this might be preferable since the field name is no longer `__root__`

Reply:
This was requested by @samuelcolvin for v1 compatibility https://github.com/pydantic/pydantic/issues/5498#issuecomment-1547625102

I will add a comment here.

Reply:
I think Samuel was more trying to say that it should be a dict, not that it should have that specific value for its key. But I might be wrong.. @samuelcolvin ?

Reply:
@lig looking back at this, I think that when Samuel says `dict` he means `model.dict()` not `dict(model)`. I think we shouldn't override the BaseModel implementation of `__iter__` here, but _should_ override `def dict()`

Reply:
Suggestion:
```python
@deprecated('The `dict` method is deprecated; use `model_dump` instead.')
    def dict(
        self,
        *,
        include: IncEx = None,
        exclude: IncEx = None,
        by_alias: bool = False,
        exclude_unset: bool = False,
        exclude_defaults: bool = False,
        exclude_none: bool = False,
    ) -> typing.Dict[str, Any]:  # noqa UP006
        warnings.warn('The `dict` method is deprecated; use `model_dump` instead.', DeprecationWarning)
        value = self.model_dump(
            include=include,
            exclude=exclude,
            by_alias=by_alias,
            exclude_unset=exclude_unset,
            exclude_defaults=exclude_defaults,
            exclude_none=exclude_none,
        )
        return {'__root__': value}
```

Reply:
this is definitely going to have breaking incompatibilities with v1 because you won't get RootModel's serialized as `{'__root__': ...}` as fields due to the use of `self.model_dump`. It would need to be an argument to `model_dump` or something for that to work. But.. I think that's a bad idea and people should just update their usage of `RootModel`.

Reply:
That's not to say that we shouldn't make this change though, if someone was calling `.dict()` on the root model I think this is reasonable.

Reply:
I discussed this with Samuel, and we agreed it makes sense to just not attempt to be backwards compatible with v1 like this because of the inconsistency that it wont' do the right thing when the rootmodel is a field.

I added a test showing how you can re-achieve this functionality by adding a mode='wrap' field serializer. I think we can add that to the documentation.